### ISSUE TITLE
run-pass-valgrind tests fail when valgrind doesn't exist

### ISSUE THREAD

**Issue by brson** [rust-lang/rust#18588]
I get `task '<main>' panicked at 'Can't find Valgrind to run Valgrind tests', /opt/dev/rust3/src/compiletest/compiletest.rs:44` when running the test suite now. These tests should probably be disabled when configure doesn't find valgrind.

**Comment by emanueLczirai** [comment]
An easy fix(unless I'm missing something, of course) would be  adding the the following lines after this one: https://github.com/rust-lang/rust/blob/81504f211c36afb69fa2a13d475c45e65d7b2a31/configure#L642

Looks like this(first line already existing):

```bash
probe CFG_VALGRIND         valgrind
if [ -z "$CFG_VALGRIND" ]
then
    if [ ! -z "$CFG_ENABLE_VALGRIND" ] || [ -z "$CFG_DISABLE_VALGRIND_RPASS" ]
    then
      err "Valgrind not found, but wanted. You may want to add --disable-valgrind-rpass"                      
    fi
fi
```
This will yell by default(./configure without params) if valgrind isn't found, unless you specifically add `--disable-valgrind-rpass`.


However, the reason why I didn't make a PR is because this code won't work in the case that the default values will change. I am speaking of these values from here: https://github.com/rust-lang/rust/blob/81504f211c36afb69fa2a13d475c45e65d7b2a31/configure#L501
reproduced here, these values I meant(0,0,1):

```bash
opt valgrind 0 "run tests with valgrind (memcheck by default)"
opt helgrind 0 "run tests with helgrind instead of memcheck"
opt valgrind-rpass 1 "run rpass-valgrind tests with valgrind"
```

adding this code within the `if [ -z "$CFG_VALGRIND" ]` will show what I mean when you test specific --enable and --disable  parameters:
```
  echo "~$CFG_ENABLE_VALGRIND~$CFG_ENABLE_VALGRIND_RPASS~$CFG_DISABLE_VALGRIND_RPASS~$CFG_DISABLE_VALGRIND"
```

Basically, if the default value for say `valgrind` is 0, then `--enable-valgrind` will set `$CFG_ENABLE_VALGRIND` but, passing `--disable-valgrind` will not set `$CFG_DISABLE_VALGRIND`.
So, if anyone will change that default valgrind value to 1, or more likely the default value `valgrind-rpass` from `1` to `0`  then the above code will have to be changed from `[ -z "$CFG_DISABLE_VALGRIND_RPASS" ]` to `[ ! -z "$CFG_ENABLE_VALGRIND_RPASS" ]`. 

 I figure this is done this way to detect if the value was specifically requested/set by the user(via a ./configure parameter), instead of the default one(./configure without params). 
 So no variable will be set($CFG_ENABLE_X / $CFG_DISABLE_X) unless the user specifically requested it via ./configure --enable_X or ./configure --disable_X .

**Comment by brson** [comment]
@emanueLczirai the problem you've discovered is an unfortunate limitation of the configure system. Writing conditionals about configuration correctly requires knowing the default value.

**Comment by tamird** [comment]
Looks like this should have been closed by #24859. @brson?

**Comment by alexcrichton** [comment]
Indeed!
